File: solutions/ruby/transpose/1/transpose.rb
# frozen_string_literal: true

# module Transpose
module Transpose
  def self.transpose(str)
    [].tap do |result|
      str.split("\n").each_with_index do |line, padding|
        line.each_char.with_index do |ch, i|
          result[i] ||= ''
          pad = [padding - result[i].length, 0].max
          result[i] += ' ' * pad + ch
        end
      end
    end.join("\n")
  end
end

File: solutions/clojure/transpose/1/src/transpose.clj
(ns transpose
  (:require [clojure.string :as st]))

(defn transpose [s]
  (let [rows (st/split-lines s)]
    (st/join \newline
             (map (fn [col]
                    (->> rows
                         reverse
                         (drop-while #(<= (count %) col))
                         reverse
                         (map #(get % col \space))
                         (apply str)))
                  (range (apply max (map count rows)))))))
